api/v1alpha1/clusterconfig_types.go

```diff
@@ -44,198 +41,181 @@ var (
+	return dockerClusterConfigVariableSchema
 }
 
-// GenericClusterConfig defines the generic cluster configdesired.
-type GenericClusterConfig struct {
+// DockerClusterConfigSpec defines the desired state of DockerClusterConfig.
+type DockerClusterConfigSpec struct {
 	// +optional
-	KubernetesImageRepository *KubernetesImageRepository `json:"kubernetesImageRepository,omitempty"`
+	Docker *DockerSpec `json:"docker,omitempty"`
+
+	GenericClusterConfigSpec `json:",inline"`
 
 	// +optional
-	Etcd *Etcd `json:"etcd,omitempty"`
+	ControlPlane *DockerNodeConfigSpec `json:"controlPlane,omitempty"`
+
+	// Extra Subject Alternative Names for the API Server signing cert.
+	// For the Nutanix provider, the following default SANs will always be added:
+	// - localhost
+	// - 127.0.0.1
+	// - 0.0.0.0
+	// - host.docker.internal
+	// +kubebuilder:validation:UniqueItems=true
+	// +kubebuilder:validation:items:Pattern=`^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$`
+	// +optional
+	ExtraAPIServerCertSANs []string `json:"extraAPIServerCertSANs,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// NutanixClusterConfig is the Schema for the nutanixclusterconfigs API.
+type NutanixClusterConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
 
 	// +optional
-	Proxy *HTTPProxy `json:"proxy,omitempty"`
+	Spec NutanixClusterConfigSpec `json:"spec,omitempty"`
+}
 
+func (s NutanixClusterConfig) VariableSchema() clusterv1.VariableSchema { //nolint:gocritic,lll // Passed by value for no potential side-effect.
+	return nutanixClusterConfigVariableSchema
+}
+
+// NutanixClusterConfigSpec defines the desired state of NutanixClusterConfig.
+type NutanixClusterConfigSpec struct {
 	// +optional
-	ExtraAPIServerCertSANs ExtraAPIServerCertSANs `json:"extraAPIServerCertSANs,omitempty"`
+	Nutanix *NutanixSpec `json:"nutanix,omitempty"`
 
+	GenericClusterConfigSpec `json:",inline"`
+
+	// +optional
+	ControlPlane *NutanixNodeConfigSpec `json:"controlPlane,omitempty"`
+
+	// Subject Alternative Names for the API Server signing cert.
+	// For the Nutanix provider, the following default SANs will always be added:
+	// - localhost
+	// - 127.0.0.1
+	// - 0.0.0.0
+	// +kubebuilder:validation:UniqueItems=true
+	// +kubebuilder:validation:items:Pattern=`^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$`
 	// +optional
-	ImageRegistries ImageRegistries `json:"imageRegistries,omitempty"`
+	ExtraAPIServerCertSANs []string `json:"extraAPIServerCertSANs,omitempty"`
+}
 
+// GenericClusterConfigSpec defines the desired state of GenericClusterConfig.
+type GenericClusterConfigSpec struct {
+	// Sets the Kubernetes image repository used for the KubeadmControlPlane.
+	// +kubebuilder:validation:Pattern=`^((?:[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*|\[(?:[a-fA-F0-9:]+)\])(:[0-9]+)?/)?[a-z0-9]+((?:[._]|__|[-]+)[a-z0-9]+)*(/[a-z0-9]+((?:[._]|__|[-]+)[a-z0-9]+)*)*$`
```

Comment:
Looks like we can remove `api/openapi/patterns/distribution.go` now.
Too bad these kubebuilder annotations can't 🪄  come from a Go var

Reply:
Yeah... it's a bit annoying tbh but we can live with it.